pkg/resource/bucket/custom_acl_code.go

```diff
@@ -0,0 +1,216 @@
+func formGrantHeader(grants []*svcsdk.Grant) string {
+	headers := []string{}
+	for _, grant := range grants {
+		if grant.Grantee == nil {
+			continue
+		}
+
+		if *grant.Grantee.Type == svcsdk.TypeGroup {
+			headers = append(headers, fmt.Sprintf(HEADER_URI_FORMAT, *grant.Grantee.URI))
+		}
+		if *grant.Grantee.Type == svcsdk.TypeCanonicalUser {
+			headers = append(headers, fmt.Sprintf(HEADER_USER_ID_FORMAT, *grant.Grantee.ID))
+		}
+	}
+	return strings.Join(headers, ",")
+}
+
+// GetHeadersFromGrants will return a list of grant headers from grants
+func GetHeadersFromGrants(
+	resp *svcsdk.GetBucketAclOutput,
+) aclGrantHeaders {
+	headers := aclGrantHeaders{
+		FullControl: formGrantHeader(getGrantsByPermission(svcsdk.PermissionFullControl, resp.Grants)),
+		Read:        formGrantHeader(getGrantsByPermission(svcsdk.PermissionRead, resp.Grants)),
+		ReadACP:     formGrantHeader(getGrantsByPermission(svcsdk.PermissionReadAcp, resp.Grants)),
+		Write:       formGrantHeader(getGrantsByPermission(svcsdk.PermissionWrite, resp.Grants)),
+		WriteACP:    formGrantHeader(getGrantsByPermission(svcsdk.PermissionWriteAcp, resp.Grants)),
+	}
+
+	return headers
+}
+
+// GetPossibleCannedACLsFromGrants will return a list of canned ACLs that match
+// the list of grants. This method will return nil if the grants did not match
+// any canned ACLs.
+func GetPossibleCannedACLsFromGrants(
+	resp *svcsdk.GetBucketAclOutput,
+) []string {
+	owner := resp.Owner
+	grants := resp.Grants
+
+	// All canned ACLs include a grant with owner full control
+	if !hasOwnerFullControl(owner, grants) {
+		return []string{}
+	}
+
+	switch len(grants) {
+	case 1:
+		return []string{CANNED_ACL_PRIVATE, CANNED_BUCKET_OWNER_READ, CANNED_BUCKET_OWNER_FULL_CONTROL}
+	case 2:
+		execTeamGrant := getGrantsByCanonicalUserID(GRANTEE_ZA_TEAM_ID, grants)
+		if grantsContainPermission(svcsdk.PermissionRead, execTeamGrant) {
+			return []string{CANNED_AWS_EXEC_READ}
+		}
+
+		allUsersGrants := getGrantsByGroupURI(GRANTEE_ALL_USERS_URI, grants)
+		if grantsContainPermission(svcsdk.PermissionRead, allUsersGrants) {
+			return []string{CANNED_PUBLIC_READ}
+		}
+
+		authenticatedUsersGrants := getGrantsByGroupURI(GRANTEE_AUTHENTICATED_USERS_URI, grants)
+		if grantsContainPermission(svcsdk.PermissionRead, authenticatedUsersGrants) {
+			return []string{CANNED_AUTHENTICATED_READ}
+		}
+	case 3:
+		logDeliveryGrants := getGrantsByGroupURI(GRANTEE_LOG_DELIVERY_URI, grants)
+		if grantsContainPermission(svcsdk.PermissionWrite, logDeliveryGrants) &&
+			grantsContainPermission(svcsdk.PermissionReadAcp, logDeliveryGrants) {
+			return []string{CANNED_LOG_DELIVERY_WRITE}
+		}
+
+		allUsersGrants := getGrantsByGroupURI(GRANTEE_ALL_USERS_URI, grants)
+		if grantsContainPermission(svcsdk.PermissionRead, allUsersGrants) &&
+			grantsContainPermission(svcsdk.PermissionWrite, allUsersGrants) {
+			return []string{CANNED_PUBLIC_READ_WRITE}
+		}
+	}
+
+	return []string{}
+}
```

Comment:
LOL, I guess this is the S3 ACL equivalent of the Sieve of Eratosthenes problem.